Migrate from `typed-builder` to `bon`

Veetaha opened the pull request.

EdJoPaTo commented on src/api_params.rs:
```diff
@@ -196,1974 +196,2069 @@ pub enum BotCommandScope {
 }
 
-#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq, Builder)]
+#[builder]
+#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
```
Mostly stating my thoughts here, not something that should be changed now:

I understand the reasoning behind having the `#[builder]` macro instead of a `#[derive(Builder)]` in order to keep it the same for structs and function and for simpler arguments. But I'm not sure whether I really like it from a perspective of using it. It still feels like something that is derived so `#[derive(Builder)]` seems intuitively more correct.

With the current `#[builder]` in its own line I would maybe just inline the `use` too: `#[bon::builder]`.

Veetaha replied:
Yeah, I understand it's conversial, but on the other hand this notation is just shorter. For example, if when you use `typed-builder` and want to add some top-level config overrides you need to both write a `derive()` and a `#[builder]` attribute e.g. 

```rust
#[derive(TypedBuilder)]
#[builder(field_defaults(...)]
struct Foo{}
```

Given that the `#[builder]` is used across many syntaxes (and there may be even more added), I find it better to keep it as a proc-macro-attribute consistently. The only downside for the users is that the `#[builder]` has to live on a separate line, which can be regarded as both a downside and an upside (it's easier to tell it apart from other harmless derives like `Debug` or `PartialEq`). Anyway, it's controversial, and I don't think `bon` would benefit from having a `derive()` variant at least at this state of my thinking to avoid the duality of syntax, and force a single consistent conventional use

EdJoPaTo reviewed (commented):
Side note: I also added your PR as a remote, and it's a bit annoying when its branch name is the same as in the base repo.

For me its also annoying when I want to create a second PR while another one is already open, so I always name them to prevent name conflicts.

And… @ayrat555 and @Veetaha: why still master, not main? :innocent:

Veetaha commented on src/api_params.rs:
```diff
@@ -196,1974 +196,2069 @@ pub enum BotCommandScope {
     #[serde(skip_serializing_if = "Option::is_none")]
-    #[builder(setter(into, strip_option), default)]
+    #[builder(into)]
     pub offset: Option<i64>,
 
     #[serde(skip_serializing_if = "Option::is_none")]
-    #[builder(setter(into, strip_option), default)]
+    #[builder(into)]
     pub limit: Option<u32>,
 
     #[serde(skip_serializing_if = "Option::is_none")]
-    #[builder(setter(into, strip_option), default)]
+    #[builder(into)]
     pub timeout: Option<u32>,
```
I'll wait for @ayrat555's response then

ayrat555 replied:
agree with @EdJoPaTo . into attributes can be removed

Veetaha replied:
Done I've removed `#[builder(into)]` from all primitive types (integers, floats, bool). The rest of the usages of `#[builder(into)]` are left on these types:

```
Box<Chat>
ChatId
FileUpload
InputMedia
Option<Animation>
Option<Audio>
Option<Birthdate>
Option<BotCommandScope>
Option<Box<Animation>>
Option<Box<Audio>>
Option<Box<Chat>>
Option<Box<ChatBackground>>
Option<Box<ChatBoostAdded>>
Option<Box<ChatShared>>
Option<Box<Contact>>
Option<Box<Dice>>
Option<Box<Document>>
Option<Box<ExternalReplyInfo>>
Option<Box<ForumTopicClosed>>
Option<Box<ForumTopicCreated>>
Option<Box<ForumTopicEdited>>
Option<Box<ForumTopicReopened>>
Option<Box<Game>>
Option<Box<GeneralForumTopicHidden>>
Option<Box<GeneralForumTopicUnhidden>>
Option<Box<InlineKeyboardMarkup>>
Option<Box<Invoice>>
Option<Box<Location>>
Option<Box<MaybeInaccessibleMessage>>
Option<Box<Message>>
Option<Box<MessageAutoDeleteTimerChanged>>
Option<Box<MessageOrigin>>
Option<Box<PaidMediaInfo>>
Option<Box<PassportData>>
Option<Box<Poll>>
Option<Box<ProximityAlertTriggered>>
Option<Box<RefundedPayment>>
Option<Box<Sticker>>
Option<Box<Story>>
Option<Box<SuccessfulPayment>>
Option<Box<TextQuote>>
Option<Box<User>>
Option<Box<UsersShared>>
Option<Box<Venue>>
Option<Box<Video>>
Option<Box<VideoChatEnded>>
Option<Box<VideoChatParticipantsInvited>>
Option<Box<VideoChatScheduled>>
Option<Box<VideoChatStarted>>
Option<Box<VideoNote>>
Option<Box<Voice>>
Option<Box<WebAppData>>
Option<BusinessIntro>
Option<BusinessLocation>
Option<BusinessOpeningHours>
Option<CallbackGame>
Option<Chat>
Option<ChatAdministratorRights>
Option<ChatId>
Option<ChatInviteLink>
Option<ChatLocation>
Option<ChatPermissions>
Option<ChatPhoto>
Option<Contact>
Option<Dice>
Option<Document>
Option<File>
Option<FileUpload>
Option<Game>
Option<Giveaway>
Option<GiveawayCompleted>
Option<GiveawayCreated>
Option<GiveawayWinners>
Option<InlineKeyboardMarkup>
Option<InlineQueryResultsButton>
Option<InputFile>
Option<InputMessageContent>
Option<Invoice>
Option<KeyboardButtonPollType>
Option<KeyboardButtonRequestChat>
Option<KeyboardButtonRequestUsers>
Option<LinkPreviewOptions>
Option<Location>
Option<LoginUrl>
Option<MaskPosition>
Option<MaybeInaccessibleMessage>
Option<MenuButton>
Option<MessageOrigin>
Option<OrderInfo>
Option<PaidMediaInfo>
Option<ParseMode>
Option<PassportFile>
Option<PhotoSize>
Option<Poll>
Option<PollType>
Option<ReplyMarkup>
Option<ReplyParameters>
Option<RevenueWithdrawalState>
Option<ShippingAddress>
Option<Sticker>
Option<StickerType>
Option<Story>
Option<String>
Option<SwitchInlineQueryChosenChat>
Option<TransactionPartner>
Option<User>
Option<Vec<AllowedUpdate>>
Option<Vec<MessageEntity>>
Option<Vec<PaidMedia>>
Option<Vec<PassportFile>>
Option<Vec<PhotoSize>>
Option<Vec<ReactionType>>
Option<Vec<ShippingOption>>
Option<Vec<String>>
Option<Vec<User>>
Option<Venue>
Option<Video>
Option<VideoNote>
Option<Voice>
Option<WebAppInfo>
Option<WriteAccessAllowed>
String
Vec<String>
```

EdJoPaTo replied:
I think there is no need for Option<…> to require an into as bon is better than types-builder in that regard. The decision should be independent of the Option here.

I also notice a bunch of `Vec<…>` things where accidental performance problems with the hidden clone on into might happen quickly. So from a gut feeling I would not into for them either as `vec![]` or `.clone()` are quite quickly used and are explicit in the user code then rather than hidden in some magic into. For String, it's a bit of a different story as the constant `"foo"` wouldn't work without into so it's likely a good idea to keep? But for message or caption it might still be a bad idea to have the into as the text could be quite big and is another performance foot gun. So maybe it should be decided per occurrence whether it's useful or not?

I also notice a bunch of Box in there and aren't sure why there are so many of them? enums? For structs there should be much benefit of using them in the first place? (probably unrelated to this PR?)

For types that don't have a from/into implemented currently the into could be removed too I guess? ChatId is likely a good case where the into is useful. But for a bigger chunk of them I currently think, it could be removed as there wouldn't be a benefit (as there is no Into/From in the first place). Currently, without a From/Into it only creates overhead for the compiler slowing everything down. (Which is already the case with typed-builder btw which would also speed up the compilation there.) So maybe removing a bigger chunk of into where there is no From/To is likely a good idea.

Veetaha replied:
As I understand `Box<>` were used to make the `Message` struct smaller in size. Optimize for the cases when it's neither an animation, nor a story, etc.

Veetaha replied:
The only `From` impls there are in this crate are for:
- `ChatId` 
- `FileUpload`
and error types (they don't participate in setters of the builders).

![image](https://github.com/user-attachments/assets/6a68491d-4304-4649-b39c-60a946f3ef8b)

I think I can just remove all `Into` from compond types. As for `String`. I think it makes sense to keep them all with `into`, even when the `String`s are used as message bodies. They can't be too big, because Telegram constraints the sizes of all messages and basically any dynamic values (including lists). Given that this library does API calls, cloning a string of under 4096 characters is not a big problem. I believe, in the contrary, people often put hardcoded string literals into messages.

Regarding `Box<>` usages. I'm not totally sure. I think `Box`-es by themselves make sense to optimize the type sizes. I'll keep an `into` on them to keep the syntax shorter for the users

Veetaha replied:
I applied the folowing NodeJS script to the two files in the repo that use builders extensively. It removes all `#[builder(into)]` where the next line doesn't contain a `String` or `ChatId` or `FileUpload` or `Box`:
```js
#!/usr/bin/env node

import fs from 'fs';

const filePaths = [
    "src/objects.rs",
    "src/api_params.rs",
]

for (const filePath of filePaths) {
    const file = fs.readFileSync(filePath, 'utf8');
    const lines = file.split('\n');

    const linesFiltered = lines.filter((line, i) => {
        if (!line.includes("#[builder(into)]")) {
            return true;
        }

        const nextLine = lines[i + 1];

        if (nextLine === undefined) {
            return true;
        }

        const patterns = [
            /\bString\b/,
            /\bChatId\b/,
            /\bFileUpload\b/,
            /\bBox<.*?>/,
        ];

        return patterns.some(pattern => pattern.test(nextLine));
    });

    fs.writeFileSync(filePath, linesFiltered.join('\n'));
}
```

EdJoPaTo replied:
Box is used due to a recursion in the type to prevent the loop. Message uses X and X uses Message again. So yeah, into for them is definitely useful.

Veetaha replied:
This is the resulting set of types with `#[builder(into)]` applied to them:
```
Box<Chat>
ChatId
FileUpload
Option<Box<Animation>>
Option<Box<Audio>>
Option<Box<Chat>>
Option<Box<ChatBackground>>
Option<Box<ChatBoostAdded>>
Option<Box<ChatShared>>
Option<Box<Contact>>
Option<Box<Dice>>
Option<Box<Document>>
Option<Box<ExternalReplyInfo>>
Option<Box<ForumTopicClosed>>
Option<Box<ForumTopicCreated>>
Option<Box<ForumTopicEdited>>
Option<Box<ForumTopicReopened>>
Option<Box<Game>>
Option<Box<GeneralForumTopicHidden>>
Option<Box<GeneralForumTopicUnhidden>>
Option<Box<InlineKeyboardMarkup>>
Option<Box<Invoice>>
Option<Box<Location>>
Option<Box<MaybeInaccessibleMessage>>
Option<Box<Message>>
Option<Box<MessageAutoDeleteTimerChanged>>
Option<Box<MessageOrigin>>
Option<Box<PaidMediaInfo>>
Option<Box<PassportData>>
Option<Box<Poll>>
Option<Box<ProximityAlertTriggered>>
Option<Box<RefundedPayment>>
Option<Box<Sticker>>
Option<Box<Story>>
Option<Box<SuccessfulPayment>>
Option<Box<TextQuote>>
Option<Box<User>>
Option<Box<UsersShared>>
Option<Box<Venue>>
Option<Box<Video>>
Option<Box<VideoChatEnded>>
Option<Box<VideoChatParticipantsInvited>>
Option<Box<VideoChatScheduled>>
Option<Box<VideoChatStarted>>
Option<Box<VideoNote>>
Option<Box<Voice>>
Option<Box<WebAppData>>
Option<ChatId>
Option<FileUpload>
Option<String>
String
```

EdJoPaTo replied:
I'm curious if there is a way to ensure stuff like this with automation. Like a lint being annoyed at `impl Into` for a type without Into/From? But its only relevant for compiler performance and independent of runtime performance.

So only for missing something. `#[bon::builder(on(…))]` allows for applying into for specific types. Maybe a "has box, then into" might be neat? That would be applied to the struct and can't be forgotten on the individual values.

Veetaha replied:
There is [`dylint`](https://github.com/trailofbits/dylint) and [marker](https://github.com/rust-marker/marker) (which I even contributed to, but unfrotunatelly, priorities changed and it did have an update for very long).

Basically, there can be a set of `into` rules defined in a macro like this in a single place:

```rust
macro_rules! define_builder {
    ($($tt:tt) => {
        #[builder(
            on(String, into),
            on(ChatId, into),
            on(FileUpload, into),
            on(Box<_>, into)
        )]
    }
}

define_builder! {
    struct Foo {}
}

define_builder! {
    struct Bar {}
}
```
This way we can ensure all structs consistently use the same `into` rules. To simplify this syntax a wrapper proc macro could be used. That proc macro could also additionally validate the other aspects of the struct (and basically serve as a custom linter). E.g. it can prohibit the direct usage of `#[builder(into)]` by checking that no struct fields have such an attribute.

Veetaha replied:
The most basic thing that is available, is that there can be a small rust binary that reads all rust files and parses them with `syn`, and then visits all `struct` items and checks all attributes.

EdJoPaTo replied:
I like the idea of a macro, but that creates more overhead per struct than I like…
In order to be really useful, a proc macro would be needed which requires its own crate…
I think that's a bit over the top.

Using ´#[bon::builder(on(String, into))]` and add the ones needed for any given struct might already be a good compromise. New structs are usually copied over so that's reducing the mistakes.
Also, the full `#[bon::builder(on(Box<_>, into), on(String, into), on(ChatId, into), on(FileUpload, into))]` could be used everywhere too.

Guess defining it once per struct and not on every field is useful and simplifies the code too.

EdJoPaTo replied:
Resolving this as its quite clear when defined per element on a struct and can be easily fixed once noticed. So not a big deal.

Veetaha commented on Cargo.toml:
```diff
@@ -76,7 +76,7 @@ required-features = ["telegram-trait"]
 [dependencies]
 thiserror = "1"
-typed-builder = "0.19"
+bon = "2.0"
```
Unfortunatelly, there isn't a single good guidance on this from the cargo book. I usually use whatever pattern people already settled with. And I usually see people specifying the `major.minor` components. I think this convention is better because by setting the minimum minor version you protect your users from unintentionally using a lower minor version which might lack the feature you are using in your lib. For example, if a new attribute or behavior is released in `bon` a minor version is bumped, and anyone depending on this new feature should bump their minor version requirement.

For patch versions it usually doesn't make sense to set a minimum requirement, unless you actually depend on some particular bug fix

EdJoPaTo replied:
Sadly, the crates ecosystem is a bit strange on what a SemVer breaking change is. cargo-semver-checks is really neat but it also (intentionally) doesn't care about rust-version changes for example. Having the bug fix version as a requirement helps with cases where the rust-version requirement changes within bug fix releases which is more often the case than I like. (And I really hate crates bumping the bug fix release version for no reasons (like dependabot does) which then changes the MSRV for no reason at all, preventing the upcoming `-Zmsrv-policy` from being useful)

Also, `2.4` can also install `2.5` which might be counterintuitive.

One kind of standard I would assume is `cargo add bon` which adds `bon = "2.0.1"`.

So personally, I use `= "2"` when it's the latest anyway and recently started with specifying the actual minimal version including the bug fix to use `-Zmsrv-policy` more effectively.

Veetaha replied:
Since I [released a `2.1`](https://www.reddit.com/r/rust/comments/1f6d7vr/media_nextgen_builder_macro_bon_21_release/) version of `bon` with important compile perf. optimizations specifically for frankenstein I think it makes sense to specify the `2.1` as a requirement.

EdJoPaTo replied:
While its more performant to have a newer version, 2.0.1 would still work.

My main concern is to either fully specify it (like `cargo add` does) or just the major version. So it's not resolved with the change to `2.1`.

Veetaha replied:
Yep, `2.0.1` will work, but people who depend on `frankenstein` will notice the compile performance degradation with that older version. So to avoid that, I just fully specified `bon = "2.1.0"`

EdJoPaTo replied:
People will use the newest possible version for a library. That's what cargo does. The point of the full version specifier is to specify the minimal version required.
For example with msrv-policy people intentionally want older versions which are marked as working. So there is no point in specifying a newer version when an older one also works.

Veetaha replied:
But what if people already have an older verison (`2.0.1`) in their dependency tree already recorded in `Cargo.lock`. In this case when ugprading to the new version of `frankenstein` their `bon` dependency won't be bumped. With this requirement, `bon` version will also be automatically updated for them when they upgrade `frankenstein`.

EdJoPaTo replied:
> But what if people already have an older verison (2.0.1) in their dependency tree already recorded in Cargo.lock. In this case when ugprading to the new version of frankenstein their bon dependency won't be bumped.

`cargo update` will just bump everything (with `-Zmsrv-policy` it will also honor your `rust-version` for example). So most users will not just bump frankenstein and ignore the rest. They will likely bump everything at once.

Personally I see much more benefit in having actual minimal version ([`cargo-minimal-versions`](https://github.com/taiki-e/cargo-minimal-versions) helps with verifying them). It allows for actually using the full range of possible versions for whatever reason the users have.
`-Zmsrv-policy` with the upcoming edition 2024 might be one. Bugs in newer dependency versions so pinning specific versions is something I need in other projects until they are fixed. So there are more reasons and I doubt I even know all of them.

Personal approach here: allow as much freedom as possible, so the users of a library can decide for themselves what's best for them. That's why I prefer minimal versions or a MSRV which allows for Debian for example.

`bon = "=2.0.0"` also seems to work fine so it's currently the minimal required version. Therefore, I would go with either `"2.0.0"` or `"2"`. Currently frankenstein doesnt define other minimal required versions (theree is no rust-version either) so I would go with that and just specify `"2"`.

Veetaha replied:
> So most users will not just bump frankenstein and ignore the rest

This is the thing. In my private work projects I never run `cargo update`. I only ever update dependencies one by one, because updating them all is almost guaranteed to break something. We have 1000+ dependendices in a cargo workspace of 170+ crates.

I agree, that for the projects of lesser scale it makes sense then. I'll change it to `bon = "2"`.

EdJoPaTo replied:
If updating all of them guarantees to break something, then something is fishy there. :D
SemVer issues? Might be a good idea to detangle that stuff a bit then :D

Veetaha replied:
When there are 1000+ dependencies, there is a big chance some of them mistakenly violated semver. From my experience, for example, one such unintentional patch in `chrono` that changed the MIN/MAX values of `Date` caused panics in production

EdJoPaTo replied:
Also: even with only updating frankenstein it will still work as the minimal required dependency versions are used. Which is your main goal: not to break something. It might not include the newest and shiniest features, but it works.

Veetaha commented:
> why still master, not main? 😇

Protecting the tradition 💪

Veetaha commented:
Small update. I measured the compilation performance of this repo both with typed-builder and bon. To my surprise, the difference is quite noticeable at this scale. Compiling from scratch (`cargo build`) on my devbox takes:
- `typed-builder`: 17 seconds
- `bon`: 22-23 seconds

I'm looking into how I can optimize `bon` based on this example. There is a sub-second performance win during the runtime of the macro itself in typed-builder (and I can definitely fine-tune `bon` to close that gap by optimizing its algorithms for sure), but the main difference is at the typechecking stage. `bon` uses a bit more traits and type indirection than `typed-builder` does. I'm currently working on a version of `bon` that avoids using traits and uses a simpler model of the generated as much as possible.

Veetaha commented:
Success! Indeed the trait abstraction in bon were causing the difficulty for the compiler. I managed to bring the type checking of the code generated by `bon` to the level of `typed-builder` :tada:. This is currently in a quickly-edited version of bon with lots of junk that I'll refactor tomorrow, but it works:

Here are the new compilation timings with `bon`:
![image](https://github.com/user-attachments/assets/e74b775a-8bc2-46cf-aacc-d7d51f58978c)
Same measurements but with `typed-builder`:
![image](https://github.com/user-attachments/assets/5fe76652-5135-4484-85d3-02b9174c2184)

There is now a `0.7`s difference between the compilation perf of `bon` and `typed-builder`, which is much more acceptable (given that `bon` depends on `darling` additionally).

Veetaha commented on src/api_params.rs:
```diff
@@ -196,1974 +196,2069 @@ pub enum BotCommandScope {
 pub struct SetWebhookParams {
-    #[builder(setter(into))]
+    #[builder(into)]
     pub url: String,
```
Let's keep this change for a separate PR

EdJoPaTo replied:
That was my intention. This is only a comment to spark the idea, mainly for @ayrat555

Veetaha commented:
Hey guys, I've updated bon to `2.1` with the compile perf. fixes and this PR is now ready for the re-review. The remaining things for this PR from the review:
- I removed `into` conversions from primitive types, but I left them on composite types (struct, enums), because some of them do have useful `From<>` implementations (see the comment https://github.com/ayrat555/frankenstein/pull/196#discussion_r1739384948)
- Regarding the cloneability of builders. I guess it's not a big deal? Either way this is going to be a breaking change (because the builder API generated by `bon` has a bit different shape.

I'm looking forward to hearing about any other things you'd like to change in this PR before merging it!

EdJoPaTo commented on examples/async_reply_to_message_updates.rs:
```diff
@@ -29,10 +29,7 @@ async fn main() {
-                        .clone()
-                        .offset(update.update_id + 1)
-                        .build();
+                    update_params = update_params_builder().offset(update.update_id + 1).build();
```
Actually, that's not the point of this example. The example modifies the existing update_params with a new offset. So maybe update only offset on this line:

```rust
update_params.offset = Some(update.update_id + 1);
```

EdJoPaTo replied:
Can you merge in the master branch? Then this diff should be gone completely. Switching typed-builder to bon without any changes in examples required is probably a good sign.

ayrat555 commented:
@Veetaha hey. can you please resolve conflicts and we can merge

Veetaha commented:
Rebase is complete 👌

Note for the future: I'm planning to change the syntax for structs to a more usual `derive(bon::Builder)` (as is often requested in feedback). This won't be a breaking change, existing code will just show deprecation warnings for `#[builder]` usages on structs. I'll create a separate PR to update `frankenstein` to the `derive(bon::Builder)` syntax once this feature is ready and released in a new minor version of bon `2.2`. Anyway, this should block this PR from merging and this will be done separately

EdJoPaTo commented:
I think we should wait a bit for bon to settle. Changes will quickly create a bunch of changes in frankenstein due to its quite heavy builder usage.

I really like its benefits over typed-builder, but I would like to prevent multiple refactorings while bon is still maturing.

Also, there might be ways to improve the usage of bon even further by condensing all `into` conditions like on String into a single place which would reduce the human mistake rate of it?

Veetaha commented:
Makes sense, I'll update the PR with the `derive` syntax once that is ready.

> Also, there might be ways to improve the usage of bon even further by condensing all into conditions like on String into a single place which would reduce the human mistake rate of it?

I don't have the best way to do that in mind, because macros in rust are stateless, so there isn't a good way to apply some global config to them. I think the simplest and optimal way to do that is with a declarative macro wrapper as described in https://github.com/ayrat555/frankenstein/pull/196#discussion_r1740194474. Do you think that'll be acceptable? Note that wrapping the struct definition with a declarative macro doesn't necessarily imply a worse IDE experience (at least in Rust Analyzer). The trick is to use a `$($tt:tt)*` input that allows for malformed (incomplete) syntax trees, so if you are editing the struct, the macro still provides useful expansion and Rust Analyzer still works just like with regular code

Veetaha commented:
Hey @ayrat555, @EdJoPaTo I've updated bon to the [version `2.2.0`](https://elastio.github.io/bon/blog/bon-builder-v2-2-release) and the diff became much more negative (there is no outstanding `#[builder]` on structs now, the previous `derive(Builder)` now works).

I propose to merge this PR. See my comment https://github.com/ayrat555/frankenstein/pull/196#issuecomment-2331646477 about the improvements suggested by @EdJoPaTo I propose to do that in a separate PR.

EdJoPaTo commented:
Given that bon has a function builder… what about the `TelegramApi::send_message` (and all the others there)?

Currently, its basically `TelegramApi::method(&self, params: &MethodParams) -> MethodResponse<Bla>`. I thought about simplifying this with a macro to reduce the repetition but maybe bon can be useful here too.

Maybe something like this could be possible?
```rust
api.send_message().message("lalala").parse_mode(ParseMode::Html).send();
```

Then the api_params wouldn't even be needed by the users of this crate.
I assume this could already work with the existing `bon::builder` when moving the parameters into the trait, but that would end up as a huge duplication of code for sync/async.

Veetaha commented:
> Maybe something like this could be possible?

Yep, it's definitely possible with `bon`.

> I assume this could already work with the existing bon::builder when moving the parameters into the trait

I suppose you meant "moving the parameters into *associated methods*"? Currently, `bon` doesn't support placing `#[builder]` on traits or impl blocks for traits.

Note that the downside of this approach is that there is indeed no "parameters" struct type, so if users want to store the parameters on some other struct, they won't be able to do that.

Veetaha commented:
Here is an [example crate](https://docs.rs/remoteit-api/0.12.2/remoteit_api/struct.R3Client.html) that provides such fluent API, although it uses an older version of `bon`. The rustdoc output improved since that quite a bit

EdJoPaTo commented:
> Here is an [example crate](https://docs.rs/remoteit-api/0.12.2/remoteit_api/struct.R3Client.html) that defines such API using `bon`, although it uses an older version of `bon`. The rustdoc output improved since that version of `bon` used there quite a bit

Doesn't look like what I've envisioned :thinking: 

> Currently, `bon` doesn't support placing `#[builder]` on traits or impl blocks for traits.

That might be a dealbreaker here.

The basic idea here is to create a builder on a trait method which uses the single parameter struct as builder methods:

```rust
struct SendMessageParams {
  pub text: String,
  pub parse_mode: ParseMode,
  …
}

trait TelegramApi {
  …
  #[whatever_needed]
  fn send_message(&self, params: SendMessageParams) -> Result<MethodResponse<Message>, …> {…}
}

struct Api {…}
impl TelegramApi for Api {…}

let api = Api::new(…);

let message = api.sendMessage().text("hello!").parse_mode(ParseMode::Html).call().unwrap();
```

One idea I had is to put the Api into the params struct, so the call has a reference to that.

The params struct should continue to exist as they are needed both sync and async and code duplication should be prevented.

But I think this gets into its own discussion independent of the PR. So not relevant to be resolved before merging.

EdJoPaTo reviewed (approved):
Jut a nitpick, otherwise I think it's a good to merge!

This is definitely a breaking change with the changes of `into` but I think its worth it.

Veetaha commented:
Yeah, support for traits/trait-impls is yet an unsolved problem and requires epic design effort. I created an issue in `bon` about that https://github.com/elastio/bon/issues/123. Eventually, I hope there will be a solution
